server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -244,7 +244,7 @@ static boolean executeInternal(SearchContext searchContext) throws QueryPhaseExe
 
             CheckedConsumer<List<LeafReaderContext>, IOException> leafSorter = l -> {};
             // try to rewrite numeric or date sort to the optimized distanceFeatureQuery
-            if ((searchContext.sort() != null) && (searchContext.size() > 0) && SYS_PROP_REWRITE_SORT) {
+            if ((searchContext.sort() != null) && ((searchContext.from() + searchContext.size()) > 0) && SYS_PROP_REWRITE_SORT) {
```

Comment:
A suggestion (feel free to ignore) -- since this is getting a bit long we could pull it out into a method like `boolean runSortOptimization(...)` for clarity.

Reply:
@jtibshirani Thanks for the feedback. I have moved this check to `tryRewriteLongSort` function.